### PR TITLE
Feature: Save settings file to filesystem

### PR DESCRIPTION
**Pull request by gabryelreyes**
- Feature allows saving new configuration parameters (for example WiFi SSID & Password) to the config file
- Renamed FileReader to FileHandler, as it can also write.
- Feature required for RemoteOTA

### CODE REVIEW

File: lib/HALSim/src/FileHandler.h [hunk 1]
```diff
@@ -71,14 +71,23 @@ class FileReader : public IFileReader
...
+    FileHandler& operator=(const FileHandler& rhs); /**< Assignment of an instance. */
 };
 
 /******************************************************************************
  * Functions
  *****************************************************************************/
 
-#endif /* FILE_READER_H_ */
\ No newline at end of file
+#endif /* FILE_HANDLER_H_ */
```
Comment (LuDubies) [inline lib/HALSim/src/FileHandler.h:93]: no empty last line

File: lib/HALTarget/src/FileHandler.h [hunk 2]
```diff
@@ -71,14 +71,23 @@ class FileReader : public IFileReader
...
+    FileHandler& operator=(const FileHandler& rhs); /**< Assignment of an instance. */
 };
 
 /******************************************************************************
  * Functions
  *****************************************************************************/
 
-#endif /* FILE_READER_H_ */
\ No newline at end of file
+#endif /* FILE_HANDLER_H_ */
```
Comment (LuDubies) [inline lib/HALTarget/src/FileHandler.h:93]: no empty line

**Review by LuDubies — APPROVED** [review]

File: lib/ConvoyLeader/src/App.cpp [hunk 3]
```diff
@@ -141,6 +141,7 @@ void App::setup()
             robotName.replace(":", "");
 
             settings.setRobotName(robotName);
+            settings.saveConfigurationFile(board.getConfigFilePath());
```
Comment (BlueAndi) [inline lib/ConvoyLeader/src/App.cpp:144]: Return value not considered

File: lib/HALInterfaces/src/IFileHandler.h [hunk 4]
```diff
@@ -69,11 +69,20 @@ class IFileReader
      */
     virtual size_t readFile(const String& fileName, char* outBuffer, const uint32_t maxBufferSize) = 0;
 
+    /**
+     * Write a file to the filesystem.
+     * @param[in] fileName Name of the file to write. Name must be an absolute path.
+     * @param[in] buffer Buffer to write file from.
+     * @param[in] bufferSize Number of bytes in the buffer.
+     * @returns number of bytes written.
```
Comment (BlueAndi) [inline lib/HALInterfaces/src/IFileHandler.h:77]: Please be aware of empty lines between description, parameters and return value.
Please start e.g. "number of bytes ..." in uppercase.

This is comment is valid for all other places as well.

File: lib/HALInterfaces/src/IFileHandler.h [hunk 5]
```diff
@@ -69,11 +69,20 @@ class IFileReader
      */
     virtual size_t readFile(const String& fileName, char* outBuffer, const uint32_t maxBufferSize) = 0;
```
Comment (BlueAndi) [inline lib/HALInterfaces/src/IFileHandler.h:70]: Could the method be const?

File: lib/HALSim/src/FileHandler.cpp [hunk 6]
```diff
@@ -94,6 +94,35 @@ size_t FileReader::readFile(const String& fileName, char* outBuffer, const uint3
...
+
+    if (nullptr == file)
+    {
+        LOG_ERROR("Failed to open file \"%s\".", fileName.c_str());
+    }
+    else
+    {
+        writtenBytes = fwrite(buffer, sizeof(char), bufferSize, file);
+
+        if (ferror(file) != 0)
```
Comment (BlueAndi) [inline lib/HALSim/src/FileHandler.cpp:110]: yoda style

File: lib/Service/src/SettingsHandler.cpp [hunk 7]
```diff
@@ -68,7 +68,7 @@ bool SettingsHandler::loadConfigurationFile(const String& filename)
     StaticJsonDocument<maxBufferSize> doc;
     char                              buffer[maxBufferSize];
 
-    if (0U == m_fileReader.readFile(filename, buffer, maxBufferSize))
+    if (0U == m_fileHandler.readFile(filename, buffer, maxBufferSize))
```
Comment (BlueAndi) [inline lib/Service/src/SettingsHandler.cpp:71]: At a lot of places the "U" is not available. Here it is of course. But we should use it consistent in many other places!

File: lib/Service/src/SettingsHandler.cpp [hunk 8]
```diff
@@ -113,6 +117,67 @@ bool SettingsHandler::loadConfigurationFile(const String& filename)
                 m_mqttPort = jsonMqttPort.as<uint16_t>();
             }
 
+            if (false == jsonApSSID.isNull())
+            {
+                m_apSSID = jsonApSSID.as<const char*>();
```
Comment (BlueAndi) [inline lib/Service/src/SettingsHandler.cpp:122]: const makes no sense ... same in the other assignements.

**Review by BlueAndi — CHANGES_REQUESTED** [review]

File: lib/Service/src/SettingsHandler.cpp (hunk 8 above)
Comment (gabryelreyes) [reply lib/Service/src/SettingsHandler.cpp:122]: ArduinoJSON does not support `jsonObject.as<char*>`, that is why all assignements are const

File: lib/Service/src/SettingsHandler.cpp (hunk 7 above)
Comment (gabryelreyes) [reply lib/Service/src/SettingsHandler.cpp:71]: Do you mean that I should remove the "U" in this line, or add it to all places where it is missing?

File: lib/Service/src/SettingsHandler.cpp (hunk 8 above)
Comment (BlueAndi) [reply lib/Service/src/SettingsHandler.cpp:122]: Oh, I see. Didn't expect that.

**Review by BlueAndi — APPROVED** [review]

**Pull request merged**